File: src/examples/protect-list/post-filter-coarse.rb
require 'kessel-sdk'

include Kessel::RBAC::V2
include Kessel::Inventory::V1beta2

# region setup
# Initialize Kessel client (see "Protect an endpoint" guide for full setup)
client = KesselInventoryService::ClientBuilder.new("localhost:9000")
                                              .insecure
                                              .build
# endregion

# region filter-by-workspace
# Filter integrations by checking workspace permissions.
#
# Post-filtering (coarse): Check workspace-level permissions for unique workspaces.
def filter_by_workspace_permission(integrations, client, user_id, permission)
  # Step 1: Get unique workspace IDs
  workspace_ids = integrations.map { |i| i[:workspace_id] }.uniq

  # Step 2: Check workspace permissions with CheckBulk
  items = workspace_ids.map do |ws_id|
    CheckBulkRequestItem.new(
      object: workspace_resource(ws_id),
      relation: permission,
      subject: principal_subject(user_id, "redhat")
    )
  end

  request = CheckBulkRequest.new(items: items)
  response = client.check_bulk(request)

  # Step 3: Build set of allowed workspaces
  allowed_workspaces = Set.new
  response.pairs.each_with_index do |pair, index|
    if pair.item&.allowed == Allowed::ALLOWED_TRUE
      allowed_workspaces << workspace_ids[index]
    end
  end

  # Step 4: Filter resources by allowed workspaces
  integrations.select { |i| allowed_workspaces.include?(i[:workspace_id]) }
end
# endregion

# region full-example
# List integrations the user can access.
#
# Post-filtering (coarse): Query all integrations, then check workspace permissions.
def list_integrations(db, client, user_id)
  # Step 1: Query database
  all_integrations = db.exec("SELECT * FROM integrations WHERE status = 'active'").to_a

  # Step 2: Filter by workspace permission
  filter_by_workspace_permission(
    all_integrations,
    client,
    user_id,
    "myservice_integration_view"
  )
end
# endregion

File: src/examples/protect-list/post-filter-coarse.ts
import { ClientBuilder } from "@project-kessel/kessel-sdk/kessel/inventory/v1beta2";
import { CheckBulkRequestItem } from "@project-kessel/kessel-sdk/kessel/inventory/v1beta2/check_bulk_request";
import { CheckBulkRequest } from "@project-kessel/kessel-sdk/kessel/inventory/v1beta2/check_bulk_request";
import { Allowed } from "@project-kessel/kessel-sdk/kessel/inventory/v1beta2/allowed";
import { principalSubject, workspaceResource } from "@project-kessel/kessel-sdk/kessel/rbac/v2";

//#region setup
// Initialize Kessel client (see "Protect an endpoint" guide for full setup)
const client = new ClientBuilder("localhost:9000")
  .insecure()
  .buildAsync();
//#endregion

//#region filter-by-workspace
/**
 * Filter integrations by checking workspace permissions.
 *
 * Post-filtering (coarse): Check workspace-level permissions for unique workspaces.
 */
async function filterByWorkspacePermission(
  integrations: Integration[],
  userId: string,
  permission: string
): Promise<Integration[]> {
  // Step 1: Get unique workspace IDs
  const workspaceIds = Array.from(new Set(integrations.map(i => i.workspaceId)));

  // Step 2: Check workspace permissions with CheckBulk
  const items: CheckBulkRequestItem[] = workspaceIds.map(wsId => ({
    object: workspaceResource(wsId),
    relation: permission,
    subject: principalSubject(userId, "redhat"),
  }));

  const request: CheckBulkRequest = { items };
  const response = await client.checkBulk(request);

  // Step 3: Build set of allowed workspaces
  const allowedWorkspaces = new Set<string>();
  response.pairs?.forEach((pair, index) => {
    if (pair.item?.allowed === Allowed.ALLOWED_TRUE) {
      allowedWorkspaces.add(workspaceIds[index]);
    }
  });

  // Step 4: Filter resources by allowed workspaces
  return integrations.filter(i => allowedWorkspaces.has(i.workspaceId));
}
//#endregion

//#region full-example
/**
 * List integrations the user can access.
 *
 * Post-filtering (coarse): Query all integrations, then check workspace permissions.
 */
async function listIntegrations(db: Database, userId: string): Promise<Integration[]> {
  // Step 1: Query database
  const allIntegrations = await db.query("SELECT * FROM integrations WHERE status = 'active'");

  // Step 2: Filter by workspace permission
  const accessible = await filterByWorkspacePermission(
    allIntegrations,
    userId,
    "myservice_integration_view"
  );

  return accessible;
}

// Database interface placeholder
interface Database {
  query(sql: string): Promise<Integration[]>;
}

interface Integration {
  id: string;
  name: string;
  workspaceId: string;
}
//#endregion
